### DISCUSSION TITLE
Build my own smart generator.

### DISCUSSION THREAD

**Discussion by spacecabbie** [jgyates/genmon#1182] [Questions]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).
- [X] I have included the generator make and model in my submission.

### Version

N/A, newest

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have not submitted logs

### What is/are your question(s)?

I have a genergy generator (spain) base 4 stroke no monitoring on it, electric start.

I plan to make it smart, Adding a rpm, oil, temp, amp, frequency, and voltage sensors. Wiring up the electric start and installing a auto choke.

Now I wonder if i add all these sensors to a raspberry can genmon read them directly or would i have to convert the sensor outputs to modbus first via a controller of sorts, and ifso any recomendations how to do that part ?

**Comment by jgyates** [comment]
Modbus is the only way genmon will input data at the moment. You can use serial or TCP.

There is a method of defining modbus register definitions and mapping them to functionality within genmon. That process is defined [here](https://github.com/jgyates/genmon/wiki/Appendix-N-Genmon-Supporting-Other-Controller-Types). Let me know if you have any questions on interfacing to genmon.

As for a modbus interface, I guess it depends on your level of comfort with programming. Python is a great environment and has numerous libraries that support modbus. In this situation genmon would be the modbus server/master and the the controller would be the modbus client/slave. [This is one example](https://pymodbus.readthedocs.io/en/latest/source/readme.html). Genmon uses it's own code (except for the CRC calculation) for modbus as it is a fairly simple protocol. Each sensor would map to a modbus register that genmon would poll to get updated info on each sensor. You would need to have a hardware / software environment that would read each sensor and have the system respond to each modbus register request with the mapped sensor value.

**Reply by spacecabbie** [reply]
> Modbus is the only way genmon will input data at the moment. You can use serial or TCP.
> 
> There is a method of defining modbus register definitions and mapping them to functionality within genmon. That process is defined [here](https://github.com/jgyates/genmon/wiki/Appendix-N-Genmon-Supporting-Other-Controller-Types). Let me know if you have any questions on interfacing to genmon.
> 
Perfect should be able to figure that out.

>  You would need to have a hardware / software environment that would read each sensor and have the system respond to each modbus register request with the mapped sensor value.

This would be the challenge then, I will need some software or something that reads the sensors and outputs modbus. 

My programming skills are very limited, But I never shy of a challenge so I'll have a look around thank you very much for the information.
